<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Introduction</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>



<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<h1>Introduction</h1>

<p>The tidycwl package takes the raw Common Workflow Language (CWL) workflows encoded in JSON or YAML, and turns the workflow elements into tidy data frames or structured lists. This package follows the <a href="https://principles.tidyverse.org/">tidyverse design principles</a> and can be seamlessly used together with the other packages with similar designs.</p>

<p>Let&#39;s use a real-world example to see how we can read, parse, and visualize a bioinformatics workflow with tidycwl.</p>

<pre><code class="r">library(&quot;tidycwl&quot;)
</code></pre>

<h1>Read workflow</h1>

<p>To read a CWL workflow into R, use <code>read_cwl_json()</code>, <code>read_cwl_yaml()</code>, or <code>read_cwl(format = ...)</code> depending on the workflow storage format.</p>

<pre><code class="r">flow &lt;- system.file(&quot;cwl/sbg/workflow/gatk4-wgs.json&quot;, package = &quot;tidycwl&quot;) %&gt;%
  read_cwl_json()
flow
</code></pre>

<pre><code>Name: Whole Genome Sequencing - BWA + GATK 4.0 (with Metrics) 
Class: Workflow 
CWL Version: sbg:draft-2 
</code></pre>

<p>We see the name, class (workflow or command line tool), and the version of the CWL. Currently, tidycwl supports both sbg:draft2 and v1.0 workflows. As the standard evolves, we plan to add the support for higher versions as needed.</p>

<h1>Parse workflow</h1>

<p>After reading the workflow into R, let&#39;s parse the main components from the CWL.</p>

<p>Besides the type (<code>parse_type()</code>) and metadata (<code>parse_meta()</code>), we are more than interested in the core components of a workflow, namely, the inputs, outputs, and the intermediate steps.</p>

<pre><code class="r">flow %&gt;%
  parse_inputs() %&gt;%
  names()
</code></pre>

<pre><code>[1] &quot;sbg:fileTypes&quot;        &quot;label&quot;                &quot;id&quot;                  
[4] &quot;sbg:includeInPorts&quot;   &quot;type&quot;                 &quot;description&quot;         
[7] &quot;sbg:category&quot;         &quot;sbg:toolDefaultValue&quot;
</code></pre>

<pre><code class="r">flow %&gt;%
  parse_outputs() %&gt;%
  names()
</code></pre>

<pre><code>[1] &quot;source&quot;             &quot;label&quot;              &quot;required&quot;          
[4] &quot;id&quot;                 &quot;sbg:includeInPorts&quot; &quot;type&quot;              
[7] &quot;sbg:fileTypes&quot;     
</code></pre>

<pre><code class="r">flow %&gt;%
  parse_steps() %&gt;%
  names()
</code></pre>

<pre><code>[1] &quot;sbg:x&quot;   &quot;inputs&quot;  &quot;outputs&quot; &quot;run&quot;     &quot;id&quot;      &quot;sbg:y&quot;   &quot;scatter&quot;
</code></pre>

<p>Depending on whether these components are represented as YAML/JSON dictionaries or lists in the workflow, the parsed results could be data frames or lists. This is because we want to keep the transformations for the original data minimal, at least at this stage. Plus, these results are not too useful compared to the following granular parsers.</p>

<h1>Get parameters</h1>

<p>We can use the <code>get_*_*()</code> functions to get the critical parameters, such as the ID, label, or documentation from the parsed inputs, outputs, and steps. For example, use <code>get_steps_label()</code> to get the labels of the steps in the workflow:</p>

<pre><code class="r">flow %&gt;%
  parse_steps() %&gt;%
  get_steps_label()
</code></pre>

<pre><code> [1] &quot;SBG Genome Coverage&quot;                 &quot;SBG Untar fasta&quot;                    
 [3] &quot;Sambamba Merge&quot;                      &quot;SBG FASTQ Quality Adjuster&quot;         
 [5] &quot;Tabix Index&quot;                         &quot;GATK CollectAlignmentSummaryMetrics&quot;
 [7] &quot;SBG Prepare Intervals&quot;               &quot;FastQC&quot;                             
 [9] &quot;BWA INDEX 0.7.17&quot;                    &quot;SBG Pair FASTQs by Metadata&quot;        
[11] &quot;SBG FASTA Indices&quot;                   &quot;Tabix BGZIP&quot;                        
[13] &quot;BWA MEM Bundle 0.7.17&quot;               &quot;GATK HaplotypeCaller&quot;               
[15] &quot;GATK IndexFeatureFile&quot;               &quot;GATK IndexFeatureFile&quot;              
[17] &quot;GATK IndexFeatureFile&quot;               &quot;GATK MergeVcfs&quot;                     
[19] &quot;GATK GenotypeGVCFs&quot;                  &quot;GATK MergeVcfs&quot;                     
[21] &quot;GATK ApplyBQSR&quot;                      &quot;GATK BaseRecalibrator&quot;              
</code></pre>

<h1>Get graph elements</h1>

<p>In many cases, it is useful to construct a graph with the parsed inputs, outputs, and steps from the workflow. The functions <code>get_nodes()</code> and <code>get_edges()</code> can help us tidy the graph nodes and edges into data frames. Each row represents a node or an edge, with each variable representing an attribute of the node or edge.</p>

<p>The function <code>get_graph()</code> is a wrapper which returns everything in a list:</p>

<pre><code class="r">get_graph(
  flow %&gt;% parse_inputs(),
  flow %&gt;% parse_outputs(),
  flow %&gt;% parse_steps()
) %&gt;% str()
</code></pre>

<pre><code>List of 2
 $ nodes:&#39;data.frame&#39;:  36 obs. of  3 variables:
  ..$ id   : chr [1:36] &quot;intervals_file&quot; &quot;dbsnp&quot; &quot;mills&quot; &quot;fastq&quot; ...
  ..$ label: chr [1:36] &quot;Target BED&quot; &quot;dbsnp&quot; &quot;Mills&quot; &quot;Fastq&quot; ...
  ..$ group: chr [1:36] &quot;input&quot; &quot;input&quot; &quot;input&quot; &quot;input&quot; ...
 $ edges:&#39;data.frame&#39;:  43 obs. of  5 variables:
  ..$ from     : chr [1:43] &quot;SBG_FASTA_Indices&quot; &quot;Sambamba_Merge&quot; &quot;reference&quot; &quot;BWA_MEM_Bundle_0_7_17&quot; ...
  ..$ to       : chr [1:43] &quot;SBG_Genome_Coverage&quot; &quot;SBG_Genome_Coverage&quot; &quot;SBG_Untar_fasta&quot; &quot;Sambamba_Merge&quot; ...
  ..$ port_from: chr [1:43] &quot;fasta_reference&quot; &quot;merged_bam&quot; NA &quot;aligned_reads&quot; ...
  ..$ port_to  : chr [1:43] &quot;fasta&quot; &quot;bam&quot; &quot;input_tar_with_reference&quot; &quot;bams&quot; ...
  ..$ type     : chr [1:43] &quot;step_to_step&quot; &quot;step_to_step&quot; &quot;input_to_step&quot; &quot;step_to_step&quot; ...
</code></pre>

<h1>Visualize workflow</h1>

<p>With tidycwl, we can visualize the workflow graph by calling <code>visualize_graph()</code>, which is built on the <code>visNetwork</code> package with an automatic hierarchical layout:</p>

<pre><code class="r">if (rmarkdown::pandoc_available(&quot;1.12.3&quot;)) {
  get_graph(
    flow %&gt;% parse_inputs(),
    flow %&gt;% parse_outputs(),
    flow %&gt;% parse_steps()
  ) %&gt;% visualize_graph()
}
</code></pre>

<div class="figure">
<!--html_preserve--><div id="htmlwidget-452f4932c817dc064daa" style="width:100%;height:600px;" class="visNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-452f4932c817dc064daa">{"x":{"nodes":{"id":["intervals_file","dbsnp","mills","fastq","reference","known_indels_1000g_bqsr","threads","memory_per_job","Genome_coverage","alignment_summary_metrics","aligned_reads","fastqc_report","gVCF","raw_vcf","SBG_Genome_Coverage","SBG_Untar_fasta","Sambamba_Merge","SBG_FASTQ_Quality_Adjuster","Tabix_Index","GATK_CollectAlignmentSummaryMetrics","SBG_Prepare_Intervals","FastQC","BWA_INDEX_1","SBG_Pair_FASTQs_by_Metadata","SBG_FASTA_Indices","Tabix_BGZIP","BWA_MEM_Bundle_0_7_17","GATK_HaplotypeCaller","GATK_IndexFeatureFile","GATK_IndexFeatureFile_1","GATK_IndexFeatureFile_2","GATK_MergeVcfs_1","GATK_GenotypeGVCFs","GATK_MergeVcfs","GATK_ApplyBQSR","GATK_BaseRecalibrator"],"label":["Target BED","dbsnp","Mills","Fastq","Reference or TAR with BWA reference indices","Known indels 1000g bqsr","Threads","Memory Per Job","Genome coverage","Alignment Summary Metrics","Aligned Reads","FastQC Report","gVCF","Raw VCF","SBG Genome Coverage","SBG Untar fasta","Sambamba Merge","SBG FASTQ Quality Adjuster","Tabix Index","GATK CollectAlignmentSummaryMetrics","SBG Prepare Intervals","FastQC","BWA INDEX 0.7.17","SBG Pair FASTQs by Metadata","SBG FASTA Indices","Tabix BGZIP","BWA MEM Bundle 0.7.17","GATK HaplotypeCaller","GATK IndexFeatureFile","GATK IndexFeatureFile","GATK IndexFeatureFile","GATK MergeVcfs","GATK GenotypeGVCFs","GATK MergeVcfs","GATK ApplyBQSR","GATK BaseRecalibrator"],"group":["input","input","input","input","input","input","input","input","output","output","output","output","output","output","step","step","step","step","step","step","step","step","step","step","step","step","step","step","step","step","step","step","step","step","step","step"]},"edges":{"from":["SBG_FASTA_Indices","Sambamba_Merge","reference","BWA_MEM_Bundle_0_7_17","fastq","Tabix_BGZIP","SBG_FASTA_Indices","Sambamba_Merge","intervals_file","fastq","reference","SBG_FASTQ_Quality_Adjuster","SBG_Untar_fasta","GATK_MergeVcfs_1","threads","BWA_INDEX_1","SBG_Pair_FASTQs_by_Metadata","SBG_FASTA_Indices","GATK_ApplyBQSR","memory_per_job","SBG_Prepare_Intervals","GATK_IndexFeatureFile","dbsnp","mills","known_indels_1000g_bqsr","GATK_HaplotypeCaller","GATK_HaplotypeCaller","SBG_FASTA_Indices","GATK_GenotypeGVCFs","SBG_Prepare_Intervals","BWA_MEM_Bundle_0_7_17","GATK_BaseRecalibrator","SBG_FASTA_Indices","BWA_MEM_Bundle_0_7_17","GATK_IndexFeatureFile","GATK_IndexFeatureFile_2","GATK_IndexFeatureFile_1","SBG_Genome_Coverage","GATK_CollectAlignmentSummaryMetrics","Sambamba_Merge","FastQC","Tabix_Index","GATK_MergeVcfs"],"to":["SBG_Genome_Coverage","SBG_Genome_Coverage","SBG_Untar_fasta","Sambamba_Merge","SBG_FASTQ_Quality_Adjuster","Tabix_Index","GATK_CollectAlignmentSummaryMetrics","GATK_CollectAlignmentSummaryMetrics","SBG_Prepare_Intervals","FastQC","BWA_INDEX_1","SBG_Pair_FASTQs_by_Metadata","SBG_FASTA_Indices","Tabix_BGZIP","BWA_MEM_Bundle_0_7_17","BWA_MEM_Bundle_0_7_17","BWA_MEM_Bundle_0_7_17","GATK_HaplotypeCaller","GATK_HaplotypeCaller","GATK_HaplotypeCaller","GATK_HaplotypeCaller","GATK_HaplotypeCaller","GATK_IndexFeatureFile","GATK_IndexFeatureFile_1","GATK_IndexFeatureFile_2","GATK_MergeVcfs_1","GATK_GenotypeGVCFs","GATK_GenotypeGVCFs","GATK_MergeVcfs","GATK_ApplyBQSR","GATK_ApplyBQSR","GATK_ApplyBQSR","GATK_BaseRecalibrator","GATK_BaseRecalibrator","GATK_BaseRecalibrator","GATK_BaseRecalibrator","GATK_BaseRecalibrator","Genome_coverage","alignment_summary_metrics","aligned_reads","fastqc_report","gVCF","raw_vcf"],"port_from":["fasta_reference","merged_bam",null,"aligned_reads",null,"output_file","fasta_reference","merged_bam",null,null,null,"result","output_fasta","output",null,"indexed_reference","tuple_list","fasta_reference","output",null,"intervals","output_with_index",null,null,null,"output_vcf","output_vcf","fasta_reference","vcf","intervals","aligned_reads","recalibration_file","fasta_reference","aligned_reads","output_with_index","output_with_index","output_with_index","summary","summary_metrics","merged_bam","report_html","indexed_file","output"],"port_to":["fasta","bam","input_tar_with_reference","bams","fastq","input_file","REFERENCE_SEQUENCE","INPUT","bed_file","input_fastq","reference","fastq_list","reference","input_file","threads","reference_index_tar","input_reads","reference","reads","memory_per_job","intervals_file","db_snp","feature_file","feature_file","feature_file","input","variants","reference","input","intervals_file","input_bam","bqsr_recal_file","reference","reads","known_sites","known_sites","known_sites",null,null,null,null,null,null],"type":["step_to_step","step_to_step","input_to_step","step_to_step","input_to_step","step_to_step","step_to_step","step_to_step","input_to_step","input_to_step","input_to_step","step_to_step","step_to_step","step_to_step","input_to_step","step_to_step","step_to_step","step_to_step","step_to_step","input_to_step","step_to_step","step_to_step","input_to_step","input_to_step","input_to_step","step_to_step","step_to_step","step_to_step","step_to_step","step_to_step","step_to_step","step_to_step","step_to_step","step_to_step","step_to_step","step_to_step","step_to_step","step_to_output","step_to_output","step_to_output","step_to_output","step_to_output","step_to_output"]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot","borderWidth":2},"manipulation":{"enabled":false},"edges":{"arrows":{"to":{"enabled":true,"scaleFactor":0.5}},"smooth":{"type":"cubicBezier","roundness":0.6}},"groups":{"input":{"color":"#C3C3C3","shadow":{"enabled":true}},"output":{"color":"#FF8F00","shadow":{"enabled":true}},"useDefaultGroups":true,"step":{"color":"#00AAA8","shadow":{"enabled":true}}},"layout":{"hierarchical":{"enabled":true,"levelSeparation":300,"direction":"LR","sortMethod":"directed"}}},"groups":["input","output","step"],"width":"100%","height":600,"idselection":{"enabled":false},"byselection":{"enabled":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)"},"evals":[],"jsHooks":[]}</script><!--/html_preserve-->
<p class="caption">plot of chunk unnamed-chunk-7</p>
</div>

<p>Users can interact with the visualization by zooming in/out and dragging the view or nodes. The graphical details can be further fine-tuned by feeding additional parameters to <code>visualize_graph()</code>.</p>

<p>The visualizations can be exported as HTML or static images (PNG/JPEG/PDF) with <code>export_html()</code> and <code>export_image()</code>.</p>

<h1>Shiny widget</h1>

<p>The workflow visualizations can be directly embedded in Shiny apps by using the Shiny widget in tidycwl. Check out the documentation for <code>render_cwl()</code> and <code>cwl_output()</code> for an example app.</p>

</body>

</html>
